src/mmds/src/token.rs

```diff
@@ -0,0 +1,541 @@
+    /// AES-GCM cipher has been reinitialized.
+    CipherReinitialized,
+    /// Failed to extract entropy from pool.
+    EntropyPool(io::Error),
+    /// Failed to extract expiry from token sequence.
+    ExpiryExtraction,
+    /// Time to live value for token is invalid.
+    InvalidTtlValue(u32),
+    /// Token serialization failed.
+    Serialization(BincodeError),
+    /// Failed to encrypt token.
+    TokenEncryption,
+}
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match &*self {
+            Error::CipherReinitialized => {
+                write!(
+                    f,
+                    "The limit of tokens generated under current MMDS token authority has been \
+                    reached. The authority entity has been reseeded and all previously created \
+                    tokens have been invalidated. Please retry the request to generate a token \
+                    under the new authority."
+                )
+            }
+            Error::EntropyPool(err) => {
+                write!(
+                    f,
+                    "Failed to extract entropy from /dev/urandom entropy pool: {}.",
+                    err
+                )
+            }
+            Error::ExpiryExtraction => write!(f, "Failed to extract expiry value from token."),
+            Error::InvalidTtlValue(value) => write!(
+                f,
+                "Invalid time to live value provided for token: {}. \
+                Please provide a value between {} and {}.",
+                value, MIN_TOKEN_TTL_SECONDS, MAX_TOKEN_TTL_SECONDS,
+            ),
+            Error::Serialization(err) => write!(f, "Bincode serialization failed: {}.", err),
+            Error::TokenEncryption => write!(f, "Failed to encrypt token."),
+        }
+    }
+}
+
+pub struct TokenAuthority {
+    cipher: aes_gcm::Aes256Gcm,
+    // Number of tokens encrypted under the current key.
+    num_encrypted_tokens: u32,
+    // Source of entropy.
+    entropy_pool: File,
+}
+
+impl TokenAuthority {
+    /// Create a new token authority entity.
+    pub fn new() -> Result<TokenAuthority, Error> {
+        let mut file = File::open(Path::new(RANDOMNESS_POOL)).map_err(Error::EntropyPool)?;
+
+        Ok(TokenAuthority {
+            cipher: TokenAuthority::create_cipher(&mut file)?,
+            num_encrypted_tokens: 0,
+            entropy_pool: file,
+        })
+    }
+
+    /// Generate encoded token string using the token time to live provided.
+    pub fn generate_token_secret(&mut self, ttl_seconds: u32) -> Result<String, Error> {
+        // Check number of tokens encrypted under the current key. We need to
+        // make sure no more than 2^32 tokens are encrypted with the same key.
+        // If this number is reached, we need to reinitialize the cipher entity
+        // and inform the guest that all previously valid tokens have been invalidated.
+        self.check_encryption_count()?;
```

Comment:
I think the previous approach (returning a token generated with the new cipher) was better. The new approach introduces a new corner case for the customer to handle.

Reply:
Previously the guest had no idea that the tokens were invalidated.
I agree that this is a new corner case to handle.
Would it be better to add a custom response header that indicates this instead? Or a new field in the body of the response that notifies the guest.

Reply:
I wouldn't add a new field in the body, as it might break guest integration. I think the response body should only contain the token if the request is successful.

Reply:
The guest as a whole won't find out that the cipher was regenerated anyway. Only the application that issues the 2^32rd request to MMDS. I don't think this makes a big difference.

Reply:
> I wouldn't add a new field in the body, as it might break guest integration. I think the response body should only contain the token if the request is successful.


Well there is no customer integration yet. Adding a JSON response would be more extensible for the future.

I agree however that this not make such a big difference, granted that we document this somewhere. Otherwise guests will not know why their tokens have been invalidated.

Reply:
Every detail regarding the behavior of MMDS will be mentioned in the documentation.